gax/src/main/java/com/google/api/gax/tracing/TracedOperationCallable.java

```diff
@@ -0,0 +1,132 @@
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.tracing;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.longrunning.OperationFuture;
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.OperationCallable;
+import com.google.api.gax.tracing.ApiTracerFactory.OperationType;
+import com.google.common.util.concurrent.MoreExecutors;
+import javax.annotation.Nonnull;
+
+/**
+ * This callable wraps a {@link OperationCallable} in a {@link ApiTracer}.
+ *
+ * <p>For internal use only.
+ */
+@InternalApi("For internal use by google-cloud-java clients only")
+public class TracedOperationCallable<RequestT, ResponseT, MetadataT>
+    extends OperationCallable<RequestT, ResponseT, MetadataT> {
+
+  private @Nonnull OperationCallable<RequestT, ResponseT, MetadataT> innerCallable;
+  private @Nonnull ApiTracerFactory tracerFactory;
+  private @Nonnull SpanName spanName;
+
+  public TracedOperationCallable(
+      @Nonnull OperationCallable<RequestT, ResponseT, MetadataT> innerCallable,
+      @Nonnull ApiTracerFactory tracerFactory,
+      @Nonnull SpanName spanName) {
+    this.innerCallable = innerCallable;
+    this.tracerFactory = tracerFactory;
+    this.spanName = spanName;
+  }
+
+  /**
+   * Wrap the initial start polling logical operations in a single parent trace using a {@link
+   * OperationType#LongRunning} {@link ApiTracer}.
+   */
+  @Override
+  public OperationFuture<ResponseT, MetadataT> futureCall(
+      RequestT request, ApiCallContext context) {
+
+    ApiTracer tracer =
```

Comment:
This is a reoccurring boilerplate code for all traced callables. Do you have any ideas how this can be "fixed"? If we were not trying to catch exception and do `finisher.onFailure()` explicitly, it would be possible to put pre and post-processing of the actual `innerCallable` in two methods, and just call them. Maybe with finally block:
```java
ApiTracer tracer = preProcess(context);
OperationFuture<ResponseT, MetadataT> future;
try { 
    future = innerCallable.futureCall(request, context);
    return future;
} finally {
    postProcess(future, tracer); // finisher is created inside this method
}
```
The posProcess witl create the finisher and will try to applly the callback no matter what. Ideally, if something went wrong in the future from the start, the callback will be called imediately (i.e. before even returning from `addCallback()` call), making finisher report `onFalure()` from the callback itself. This  may also solve the potential risk of double reporting the failure (once in the catch on line 83 and once the actual callback).

Do the thoughts above make sense (I hope they do, but not sure)?

Reply:
I don't think this will work, if futureCall threw an exception, how will postprocess know about it?